### DISCUSSION TITLE
How to create a fork of jetbot-repo for a custom pre-built robot?

### DISCUSSION THREAD

**Discussion by Hoehlenbaer** [NVIDIA-AI-IOT/jetbot#460] [Q&A] [answered]
Hi together,

I adapted the current jetbot 4.3 image to this pre-built robot:
https://category.yahboom.net/products/jetbot?variant=31210264100948

As the yahboom robot has got some more motors, a different turning behaviour (due to the chain drive) an a non-standard interface board, I had to change a lot of python scripts and physical addresses.

Now I want to reflect these changes in a fork, so someone else is able to use it as well. I was already able to fix the docker files, but I'm a little confused how to handle the python scripts.

Looking at this code fragment of the base dockerfile, I can see that the repo folder "/opt/jetbot" will be copied from the host (?) to the docker image.

ENV JETBOT_REPO_DIR=/opt/jetbot
COPY . ${JETBOT_REPO_DIR}
RUN cd ${JETBOT_REPO_DIR} python3 setup.py install

However how do I get this folder to the host file system? The folder "/opt/jetbot " isn't part of this repo, so a simple "git clone" isn't working. 

Can anyone help?

Thanks!
Marco

**Comment by santaimpersonator** [comment, marked as answer]
@Hoehlenbaer For the newer jetpack releases, Nvidia decided to use Docker containers.  Those lines of code you found, are used to build this jetbot Python package (this repo) within the container's workspace. Previously, they [built the package directly from the repo](https://github.com/NVIDIA-AI-IOT/jetbot/wiki/Software-Setup#step-5---install-latest-software-optional). You'll find that you can modify the Python files directly and the changes will be reflected once the Docker container is rebuilt.

Take a look at my pull request here: https://github.com/NVIDIA-AI-IOT/jetbot/pull/467
- If you clone this [branch of our fork](https://github.com/sparkfun/jetbot/tree/update-SparkFun_hardware) and build the Docker container, you'll find the changes to the Python files reflected in the Docker container.

<br>
<br>
=================================================================================

**Note:** I already tried telling @TCIII that the [Python packages run from within the Docker container workspace](https://github.com/NVIDIA-AI-IOT/jetbot/issues/466#issuecomment-870358171). Good luck to anyone else who tries to help... 🤣

**Reply by Hoehlenbaer** [reply]
Hi santaimpersonator,

Does this mean, I have to copy the files via GIT CLONE, and then build the container?
In this case life's actually much easier than I thought...

**Reply by santaimpersonator** [reply]
@Hoehlenbaer That is correct.

Just make sure to stop and remove any running containers with the [disable.sh script](https://github.com/NVIDIA-AI-IOT/jetbot/blob/master/docker/disable.sh) before building you re-build the container with your changes *(as noted in [this documentation on replacing an old container](https://docs.docker.com/get-started/03_updating_app/#replace-the-old-container))*. You can find directions for building and enabling a new container in the [README.md of the `docker` folder](https://github.com/NVIDIA-AI-IOT/jetbot/tree/master/docker).

I should note in **Step 1** of that documentation, you only need to perform the first 2 parts:
- Make sure to run the `configure.sh` script as `source`:
  ```
  source configure.sh
  ```
  Otherwise, your environment variables won't be set.
- The first part of the step will disable the GUI:
  ```
  cd jetbot
  ./scripts/configure_jetson.sh
  ```
- The last two parts are already executed in these lines of the [`configure.sh` script](https://github.com/NVIDIA-AI-IOT/jetbot/blob/master/docker/configure.sh#L24-L34).

Summary:

1. Disable containers with `disable.sh` script
2. Configure environment variables with `configure.sh` script
3. (Re-)Build new containers with `build.sh` script
4. Enable the containers with `enable.sh` script

**Comment by TCIII** [comment]


**Reply by santaimpersonator** [reply]
No problem.

Performance:
- I haven't gotten to test the pre-trained models out; they don't work with the new jetpack version.
- The collision avoidance and teleoperation examples seem to run smoother than in jetpack 4.3. However, they added improvements for how the camera frames are handled since then. (**Note:** This was working with the 4GB board and running it at max power)

Modification:
You'd have to access the Docker workspace and then python folders will probably be in either the hidden `/.local/lib/` folder or the `/usr/local/lib/` folder.

However, as I mentioned before, the container is a "virtual environment". Therefore, the container resets to the initial (or clean/fresh) build state. It would be better to make the changes prior to (or within) the build process.
- If it is something you can configure through the Adafruit package, I would modify the `motor.py` and/or the `robot.py` files in this repo before building a new Docker container.

**Reply by santaimpersonator** [reply]
Also, congrats on your victory 🏆 . I hope you had fun at the event!

**Reply by santaimpersonator** [reply]
> 
> 
> @santaimpersonator,
> 
> Unfortunately the PWM switching frequency can only be modified in the Adafruit_MotorHAT library where the PWM switching frequency is adjustable.
> Yes, I had a lot of fun and met a lot of great people out at the reservoir during AVC 2013.


> 
> 
> @santaimpersonator,
> Setup.py in the jetbot directory requires the installation of Adafruit_MotorHAT library and the Adafruit-SSD1306 library, however I have yet to determine where those libraries are installed in Docker or the Containers. I have searched the IoT, but no joy. Any suggestions will be greatly appreciated.
> 
> I have inspected both of the Jetbot Containers and have found that the Jetbot Repo should be in /opt/jetbot, however there is no Jetbot directory in the opt directory.

@TCIII Try to keep the discussion in a single thread.

[As I mentioned before](https://github.com/NVIDIA-AI-IOT/jetbot/issues/466#issuecomment-870358171), you have to be in the Docker container's workspace *(i.e. you need to be in the workspace of the virtual environment provided by the Docker container)*. The easiest way to do that is by accessing the terminal through the Jupyter labs (web browser).

Please, refer to the [documentation on Docker containers here](https://www.docker.com/get-started) to get a better idea of how a Docker container operates.
